catroid/src/main/java/org/catrobat/catroid/formulaeditor/datacontainer/DataContainer.java

```diff
@@ -24,350 +24,246 @@
-	}
+	public UserList addSpriteUserList(Sprite sprite, String name) {
+		UserList list = new UserList(name);
 
-	private void resetUserLists(List<UserList> userList) {
-		spriteListBehaviour.reset(userList);
-	}
+		if (spriteListOfLists.containsKey(sprite)) {
+			spriteListOfLists.get(sprite).add(list);
+		} else {
+			spriteListOfLists.put(sprite, Collections.singletonList(list));
+		}
 
-	public UserVariable addUserBrickVariableToUserBrick(UserBrick userBrick, String userVariableName, Object value) {
-		return userBrickVariableBehaviour.add(userBrick, userVariableName, value);
+		return list;
 	}
 
-	public List<UserVariable> getOrCreateVariableListForUserBrick(UserBrick userBrick) {
-		return userBrickVariableBehaviour.getOrCreate(userBrick);
+	public UserList addProjectUserList(String name) {
+		UserList list = new UserList(name);
+		projectLists.add(list);
+		return list;
 	}
 
-	public boolean existUserVariableWithName(String name) {
-		for (UserBrick userBrick : userBrickVariables.keySet()) {
-			if (userBrickVariableBehaviour.exists(userBrick, name)) {
-				return true;
+	public void removeAllDataObjectsOfClones() {
+		for (Iterator<Map.Entry<Sprite, List<UserVariable>>> iterator = spriteVariables.entrySet().iterator();
+				iterator.hasNext(); ) {
+			if (iterator.next().getKey().isClone()) {
+				iterator.remove();
 			}
 		}
-		return false;
-	}
-
-	public UserVariable findUserBrickVariable(UserBrick userBrick, String name) {
-		return userBrickVariableBehaviour.find(userBrick, name);
-	}
-
-	public UserVariable addProjectUserVariable(String userVariableName) {
-		return projectVariableBehaviour.add(userVariableName);
-	}
-
-	public UserVariable renameProjectUserVariable(String newName, String oldName) {
-		return projectVariableBehaviour.rename(newName, oldName);
-	}
-
-	public boolean existProjectVariable(UserVariable variable) {
-		return getProjectVariables().contains(variable);
-	}
 
-	public boolean existProjectVariableWithName(String name) {
-		return projectVariableBehaviour.exists(name);
+		for (Iterator<Map.Entry<Sprite, List<UserList>>> iterator = spriteListOfLists.entrySet().iterator();
+				iterator.hasNext(); ) {
+			if (iterator.next().getKey().isClone()) {
+				iterator.remove();
+			}
+		}
 	}
 
-	public UserVariable findProjectVariable(String name) {
-		return projectVariableBehaviour.find(name);
-	}
+	public void removeAllDataObjects(Sprite sprite) {
+		spriteVariables.remove(sprite);
+		spriteListOfLists.remove(sprite);
 
-	public UserList addProjectUserList(String userListName) {
-		return projectListBehaviour.add(userListName);
+		for (UserBrick userBrick : sprite.getUserBrickList()) {
+			userBrickVariables.remove(userBrick);
+		}
 	}
 
-	public UserList renameProjectUserList(String newName, String oldName) {
-		return projectListBehaviour.rename(newName, oldName);
-	}
+	public void resetAllDataObjects() {
+		for (UserVariable var : projectVariables) {
+			var.setValue(0.0);
```

Comment:
What do you thing about using following overloaded methods? The `UserVariable`s of `UserBrick`s will need to be reset as well somewhere, right?

```java
private void resetData(UserVariable userVariable) {
	userVariable.setValue(0.0);
}

private void resetData(UserList userList) {
	userList.getList().clear();
}
```

Reply:
Alternatively, you could also add a `resetData` method to the `UserData` interface.